x-pack/plugins/monitoring/public/application/hooks/use_breadcrumbs.ts

```diff
@@ -0,0 +1,289 @@
+    breadcrumbs.push(createCrumb('#/beats', beatsLabel));
+    breadcrumbs.push(
+      createCrumb(
+        '#/beats/beats',
+        i18n.translate('xpack.monitoring.breadcrumbs.beats.instancesLabel', {
+          defaultMessage: 'Instances',
+        })
+      )
+    );
+    breadcrumbs.push(createCrumb(null, mainInstance.instance));
+  } else {
+    breadcrumbs.push(createCrumb(null, beatsLabel));
+  }
+
+  return breadcrumbs;
+}
+
+// generate Apm breadcrumbs
+function getApmBreadcrumbs(mainInstance: any) {
+  const apmLabel = i18n.translate('xpack.monitoring.breadcrumbs.apmLabel', {
+    defaultMessage: 'APM server',
+  });
+  const breadcrumbs = [];
+  if (mainInstance.instance) {
+    breadcrumbs.push(createCrumb('#/apm', apmLabel));
+    breadcrumbs.push(
+      createCrumb(
+        '#/apm/instances',
+        i18n.translate('xpack.monitoring.breadcrumbs.apm.instancesLabel', {
+          defaultMessage: 'Instances',
+        })
+      )
+    );
+    breadcrumbs.push(createCrumb(null, mainInstance.instance));
+  } else {
+    // don't link to Overview when we're possibly on Overview or its sibling tabs
+    breadcrumbs.push(createCrumb(null, apmLabel));
+  }
+  return breadcrumbs;
+}
+
+function buildBreadcrumbs(clusterName: string, mainInstance?: any | null) {
+  const homeCrumb = i18n.translate('xpack.monitoring.breadcrumbs.clustersLabel', {
+    defaultMessage: 'Clusters',
+  });
+
+  let breadcrumbs = [createCrumb('#/home', homeCrumb, 'breadcrumbClusters', true)];
+
+  if (!mainInstance?.inOverview && clusterName) {
+    breadcrumbs.push(createCrumb('#/overview', clusterName));
+  }
+
+  if (mainInstance?.inElasticsearch) {
+    breadcrumbs = breadcrumbs.concat(getElasticsearchBreadcrumbs(mainInstance));
+  }
+  if (mainInstance?.inKibana) {
+    breadcrumbs = breadcrumbs.concat(getKibanaBreadcrumbs(mainInstance));
+  }
+  if (mainInstance?.inLogstash) {
+    breadcrumbs = breadcrumbs.concat(getLogstashBreadcrumbs(mainInstance));
+  }
+  if (mainInstance?.inBeats) {
+    breadcrumbs = breadcrumbs.concat(getBeatsBreadcrumbs(mainInstance));
+  }
+  if (mainInstance?.inApm) {
+    breadcrumbs = breadcrumbs.concat(getApmBreadcrumbs(mainInstance));
+  }
+
+  return breadcrumbs;
+}
+interface BreadcrumbItem {
+  ['data-test-subj']?: string;
+  href?: string;
+  text: string;
+  ignoreGlobalState?: boolean;
+}
+
+export const useBreadcrumbs = ({ history }: { history: History }) => {
+  const chrome = useKibana().services.chrome;
+  const [breadcrumbs, setBreadcrumbs] = useState<Crumb[]>([]);
+
+  const update = useCallback(
+    (bcrumbs?: BreadcrumbItem[]) => {
+      if (!chrome) return;
+      if (!bcrumbs) {
+        const currentBreadcrumbs: Observable<any> & {
+          value?: BreadcrumbItem[];
+        } = chrome.getBreadcrumbs$()?.source;
+        if (currentBreadcrumbs && currentBreadcrumbs.value) {
+          bcrumbs = currentBreadcrumbs.value;
+        }
+      }
+      const globalStateStr = location.hash.split('?')[1];
```

Comment:
The direct location hash parsing was a little surprising to see. I was expected we'd already have some sort of kibana level breadcrumb manager that look care of this sort of thing, but maybe not?

Seems like we should probably have a test on this, if not now then before we call the react conversion "done".